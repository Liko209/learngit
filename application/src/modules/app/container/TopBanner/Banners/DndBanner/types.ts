/*
 * @Author: joy.zhang (joy.zhang@ringcentral.com)
 * @Date: 2019-07-23 01:48:27
 * Copyright Â© RingCentral. All rights reserved.
 */

type DndBannerProps = {};

type DndBannerViewProps = {
  isShow: boolean;
  handleClose: () => void;
  handleUnblock: () => void;
};

export { DndBannerProps, DndBannerViewProps };
