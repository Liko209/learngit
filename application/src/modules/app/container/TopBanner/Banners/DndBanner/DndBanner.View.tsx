/*
 * @Author: joy.zhang (joy.zhang@ringcentral.com)
 * @Date: 2019-07-23 01:48:27
 * Copyright Â© RingCentral. All rights reserved.
 */

import { observer } from 'mobx-react';
import React, { Component } from 'react';
import { JuiSnackbarContent, JuiSnackbarAction } from 'jui/components/Snackbars';
import { withTranslation, WithTranslation } from 'react-i18next';
import { ToastType } from '@/containers/ToastWrapper/Toast/types';
import { DndBannerViewProps } from './types';

type Props = DndBannerViewProps & WithTranslation;

@observer
class DndBannerViewComponent extends Component<Props> {
  render() {
    const { t, isShow, handleClose, handleUnblock } = this.props;
    return isShow && (
      <JuiSnackbarContent
        data-test-automation-id="dnd-top-banner"
        type={ToastType.ERROR}
        message={t('presence.prompt.topBannerInfoWhenDnd')}
        messageAlign="center"
        fullWidth
        action={[
          <JuiSnackbarAction
            key="unblock"
            onClick={handleUnblock}
            data-test-automation-id="dnd-top-banner-unblock"
          >
            {t('common.button.turnOff')}
          </JuiSnackbarAction>,
          <JuiSnackbarAction
            key="close"
            variant="icon"
            onClick={handleClose}
            data-test-automation-id="dnd-top-banner-close"
          >
            close
          </JuiSnackbarAction>,
        ]}
      />
    );
  }
}

const DndBannerView = withTranslation('translations')(DndBannerViewComponent);

export { DndBannerView };
