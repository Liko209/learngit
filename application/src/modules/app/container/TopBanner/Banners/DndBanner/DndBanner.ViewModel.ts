/*
 * @Author: joy.zhang (joy.zhang@ringcentral.com)
 * @Date: 2019-07-23 01:48:27
 * Copyright Â© RingCentral. All rights reserved.
 */
import { ServiceLoader, ServiceConfig } from 'sdk/module/serviceLoader';
import { action, observable, computed } from 'mobx';
import { StoreViewModel } from '@/store/ViewModel';
import { getGlobalValue, getPresence } from '@/store/utils';
import { GLOBAL_KEYS } from '@/store/constants';
import { analyticsCollector } from '@/AnalyticsCollector';
import { PRESENCE } from 'sdk/module/presence/constant';
import { catchError } from '@/common/catchError';
import { PresenceService } from 'sdk/module/presence';
import { DndBannerProps, DndBannerViewProps } from './types';

class DndBannerViewModel extends StoreViewModel<DndBannerProps> implements DndBannerViewProps {
  private _presenceService = ServiceLoader.getInstance<PresenceService>(
    ServiceConfig.PRESENCE_SERVICE,
  );

  @observable
  isClosed = false;

  @computed
  get currentUserId() {
    return getGlobalValue(GLOBAL_KEYS.CURRENT_USER_ID);
  }

  @computed
  get isDND() {
    const presence = getPresence(this.currentUserId);
    return presence === PRESENCE.DND;
  }

  @computed
  get isShow() {
    return !this.isClosed && this.isDND;
  }

  @catchError.flash({
    network: 'presence.prompt.updatePresenceFailedForNetworkIssue',
    server: 'presence.prompt.updatePresenceFailedForServerIssue',
  })
  @action
  handleUnblock = async () => {
    const toPresence = PRESENCE.AVAILABLE;
    analyticsCollector.setPresence(toPresence);
    await this._presenceService.setPresence(toPresence);
  };

  @action
  handleClose = () => {
    this.isClosed = true;
  };
}

export { DndBannerViewModel };
