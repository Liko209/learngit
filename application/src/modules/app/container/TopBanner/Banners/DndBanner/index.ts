/*
 * @Author: joy.zhang (joy.zhang@ringcentral.com)
 * @Date: 2019-07-23 01:48:27
 * Copyright Â© RingCentral. All rights reserved.
 */

export * from './DndBanner';
