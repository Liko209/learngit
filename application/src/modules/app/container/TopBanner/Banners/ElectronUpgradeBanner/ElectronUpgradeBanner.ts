/*
 * @Author: Chris Zhan (chris.zhan@ringcentral.com)
 * @Date: 2019-04-30 15:37:35
 * Copyright Â© RingCentral. All rights reserved.
 */

import { buildContainer } from '@/base';
import { ElectronUpgradeBannerView } from './ElectronUpgradeBanner.View';
import { ElectronUpgradeBannerViewModel } from './ElectronUpgradeBanner.ViewModel';
import { ElectronUpgradeBannerProps } from './types';

const ElectronUpgradeBanner = buildContainer<ElectronUpgradeBannerProps>({
  ViewModel: ElectronUpgradeBannerViewModel,
  View: ElectronUpgradeBannerView,
});

export { ElectronUpgradeBanner, ElectronUpgradeBannerProps };
