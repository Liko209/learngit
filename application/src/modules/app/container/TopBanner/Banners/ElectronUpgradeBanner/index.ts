/*
 * @Author: Chris Zhan (chris.zhan@ringcentral.com)
 * @Date: 2019-04-30 15:39:23
 * Copyright Â© RingCentral. All rights reserved.
 */

export * from './ElectronUpgradeBanner';
