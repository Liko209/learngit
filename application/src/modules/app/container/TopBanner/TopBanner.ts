/*
 * @Author: Lip Wang (lip.wang@ringcentral.com)
 * @Date: 2018-12-04 15:57:44
 * Copyright Â© RingCentral. All rights reserved.
 */
import { buildContainer } from '@/base';
import { TopBannerView } from './TopBanner.View';
import { TopBannerViewModel } from './TopBanner.ViewModel';

const TopBanner = buildContainer({
  ViewModel: TopBannerViewModel,
  View: TopBannerView,
});

export { TopBanner };
