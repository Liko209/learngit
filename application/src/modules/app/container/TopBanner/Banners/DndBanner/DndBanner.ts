/*
 * @Author: joy.zhang (joy.zhang@ringcentral.com)
 * @Date: 2019-07-23 01:48:27
 * Copyright Â© RingCentral. All rights reserved.
 */

import { buildContainer } from '@/base';
import { DndBannerView } from './DndBanner.View';
import { DndBannerViewModel } from './DndBanner.ViewModel';
import { DndBannerProps } from './types';

const DndBanner = buildContainer<DndBannerProps>({
  View: DndBannerView,
  ViewModel: DndBannerViewModel,
});

export { DndBanner };
