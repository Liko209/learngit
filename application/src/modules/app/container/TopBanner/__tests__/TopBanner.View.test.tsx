/*
 * @Author: Lip Wang (lip.wang@ringcentral.com)
 * @Date: 2018-12-06 10:32:47
 * Copyright Â© RingCentral. All rights reserved.
 */
import React from 'react';
import { shallow, mount } from 'enzyme';
import { TopBannerView } from '../TopBanner.View';
import { ElectronUpgradeBanner } from '../Banners/ElectronUpgradeBanner/ElectronUpgradeBanner';
import { NetworkBanner } from '../Banners/NetworkBanner';
import { DndBanner } from '../Banners/DndBanner';
import { container } from 'framework';

jest.spyOn(container, 'get').mockReturnValue({
  topBanners: [],
});

describe('TopBannerView', () => {
  describe('render', () => {
    it('should not render networkBannerView', () => {
      const wrapper = shallow(<TopBannerView />);
      expect(wrapper.find(NetworkBanner).length).toBe(0);
    });

    it('should render banners defined in data', () => {
      jest.spyOn(container, 'get').mockReturnValue({
        topBanners: [
          {
            priority: 200,
            Component: ElectronUpgradeBanner,
            props: {},
            isShow: true,
          },
        ],
      });
      const wrapper = shallow(<TopBannerView />);
      expect(wrapper.find(ElectronUpgradeBanner).length).toBe(1);
    });

    it('should not render banners defined in data if isShow is false', () => {
      jest.spyOn(container, 'get').mockReturnValue({
        topBanners: [
          {
            priority: 200,
            Component: ElectronUpgradeBanner,
            props: {},
            isShow: false,
          },
        ],
      });
      const wrapper = shallow(<TopBannerView />);
      expect(wrapper.find(ElectronUpgradeBanner).length).toBe(0);
    });

    it('should not display banner defined in data if priority is less than other banner [JPT-2531]', () => {
      jest.spyOn(container, 'get').mockReturnValue({
        topBanners: [
          {
            priority: 200,
            Component: DndBanner,
            props: {},
            isShow: true,
          },
          {
            priority: 100,
            Component: NetworkBanner,
            props: {},
            isShow: true,
          },
        ],
      });
      const wrapper = mount(<TopBannerView />);
      expect(wrapper).toMatchSnapshot();
    });
  });
});
