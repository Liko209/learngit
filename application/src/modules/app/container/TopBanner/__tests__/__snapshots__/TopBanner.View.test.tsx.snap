// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopBannerView render should not display banner defined in data if priority is less than other banner [JPT-2531] 1`] = `
.c0 {
  position: relative;
  overflow: hidden;
}

.c0 > :not(:first-child) {
  display: none;
}

<TopBannerView>
  <styled.div>
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bdVaJa",
            "isStatic": true,
            "lastClassName": "c0",
            "rules": Array [
              "
  position: relative;
  overflow: hidden;

  > :not(:first-child) {
    display: none;
  }
",
            ],
          },
          "displayName": "styled.div",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bdVaJa",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
    >
      <div
        className="c0"
      >
        <Container(NetworkBannerViewComponent)
          key="100"
        >
          <NetworkBannerViewComponent
            isShow={false}
            t={[Function]}
          />
        </Container(NetworkBannerViewComponent)>
        <Container(DndBannerViewComponent)
          key="200"
        >
          <DndBannerViewComponent
            currentUserId={0}
            handleClose={[Function]}
            handleUnblock={[Function]}
            isClosed={false}
            isDND={false}
            isShow={false}
            t={[Function]}
          />
        </Container(DndBannerViewComponent)>
      </div>
    </StyledComponent>
  </styled.div>
</TopBannerView>
`;
